File: problems/数学与位运算/1269.停在原地的方案数.py
#!/usr/bin/env python3
"""
1269. 停在原地的方案数
"""
def num_ways(steps: int, arr_len: int) -> int:
    MOD = 10**9 + 7
    max_pos = min(steps // 2, arr_len - 1)
    dp = [0] * (max_pos + 1)
    dp[0] = 1
    for _ in range(1, steps + 1):
        new_dp = [0] * (max_pos + 1)
        for pos in range(max_pos + 1):
            new_dp[pos] = (new_dp[pos] + dp[pos]) % MOD
            if pos > 0:
                new_dp[pos] = (new_dp[pos] + dp[pos-1]) % MOD
            if pos < max_pos:
                new_dp[pos] = (new_dp[pos] + dp[pos+1]) % MOD
        dp = new_dp
    return dp[0]

if __name__ == "__main__":
    # 测试用例
    print(num_ways(3, 2))   # 预期: 3
    print(num_ways(2, 4))   # 预期: 2
    print(num_ways(4, 2))   # 预期: 8

File: problems/数学与位运算/1269.停在原地的方案数.cpp
#include <iostream>
#include <vector>
using namespace std;

int numWays(int steps, int arrLen) {
    int maxPos = min(steps/2, arrLen-1);
    vector<int> dp(maxPos+1, 0), ndp(maxPos+1, 0);
    dp[0] = 1;
    const int MOD = 1e9+7;
    for (int s = 1; s <= steps; s++) {
        fill(ndp.begin(), ndp.end(), 0);
        for (int p = 0; p <= maxPos; p++) {
            ndp[p] = (ndp[p] + dp[p]) % MOD;
            if (p > 0) ndp[p] = (ndp[p] + dp[p-1]) % MOD;
            if (p < maxPos) ndp[p] = (ndp[p] + dp[p+1]) % MOD;
        }
        dp.swap(ndp);
    }
    return dp[0];
}

int main() {
    // 测试用例
    cout << numWays(3, 2) << endl;  // 预期: 3
    cout << numWays(2, 4) << endl;  // 预期: 2
    cout << numWays(4, 2) << endl;  // 预期: 8
    return 0;
}
